### PR TITLE
Full interceptor support 

### PR DESCRIPTION
**Pull request by blackcat**

### CODE REVIEW

**Comment by mslipper** [comment]
@blackcat ping me when you're ready for review here.

**Comment by blackcat** [comment]
@mslipper hi! Now its done.

**Comment by mslipper** [comment]
Will review tonight.

File: src/main/java/com/googlecode/jsonrpc4j/JsonRpcBasicServer.java [hunk 1]
```diff
@@ -849,8 +848,12 @@ private JsonError writeAndFlushValueError(OutputStream output, ErrorObjectWithJs
 	 * @param value  the value to write
 	 * @throws IOException on error
 	 */
-	private void writeAndFlushValue(OutputStream output, Object value) throws IOException {
+	private void writeAndFlushValue(OutputStream output, ObjectNode value) throws IOException {
```
Comment (mslipper) [inline src/main/java/com/googlecode/jsonrpc4j/JsonRpcBasicServer.java]: Do you know for sure that this will always be of type `ObjectNode`?

File: src/main/java/com/googlecode/jsonrpc4j/JsonRpcInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+ * Interceptors could throw {@link RuntimeException}, in this case JSON RPC standard error will be shown, except preHandleJson.
+ * In case preHandleJson throws exception, it will generate HTTP code 400 "Bad Request" with empty body. So be careful.
+ * @author pavyazankin
+ * @since 21/09/2017
+ */
+public interface JsonRpcInterceptor {
+
+    /**
+     * Method to intercept raw JSON RPC json objects. Don't throw exceptions here.
+     * In case preHandleJson throws exception, it will be HTTP code 400 "Bad Request" with empty body. So be careful.
```
Comment (mslipper) [inline src/main/java/com/googlecode/jsonrpc4j/JsonRpcInterceptor.java]: Is it possible for you to handle exceptions from this method? People will end up throwing exceptions in this method; could be confusing for users if the response content changes depending on where the exception is thrown.

File: src/test/java/com/googlecode/jsonrpc4j/server/JsonRpcServerTest.java [hunk 3]
```diff
@@ -16,241 +18,347 @@
...
-		Assert.assertEquals("gzip", response.getHeader(CONTENT_ENCODING));
-	}
-	
-	// Service and service interfaces used in test
-	
-	public interface ServiceInterface {
-		String testMethod(String param1);
-	}
+
+    @Mock(type = MockType.NICE)
```
Comment (mslipper) [inline src/test/java/com/googlecode/jsonrpc4j/server/JsonRpcServerTest.java]: I know that the indentation was incorrect in this file, however would you mind changing it back to the way it was before? It's tough to track exactly what changed when so many unrelated lines change in a diff.

File: src/main/java/com/googlecode/jsonrpc4j/JsonRpcBasicServer.java (hunk 1 above)
Comment (blackcat) [reply src/main/java/com/googlecode/jsonrpc4j/JsonRpcBasicServer.java]: Yes. I have checked usages before change type. There are only two usages and it is ObjectNode in both cases.

Comment (mslipper) [reply src/main/java/com/googlecode/jsonrpc4j/JsonRpcBasicServer.java]: 👍

File: src/main/java/com/googlecode/jsonrpc4j/JsonRpcInterceptor.java (hunk 2 above)
Comment (blackcat) [reply src/main/java/com/googlecode/jsonrpc4j/JsonRpcInterceptor.java]: In progress...

Comment (blackcat) [reply src/main/java/com/googlecode/jsonrpc4j/JsonRpcInterceptor.java]: It's possible to handle but process of handling will differ from other methods. Problem is that we don't have any request details that need for method  handleError(OutputStream output, Object id, String jsonRpc, AMethodWithItsArgs methodArgs, Throwable e) [375]
We didn't parse jsonRpc request yet. We know its JSON, but what is in it? In this situation I don't even know are there any jsonRpc version, correct method name, arguments. And if I'll try to parse them I could fail with other exception...
Document differences in behaviour in this case is the best choice for us.
